File: project/backend-services/notification-dispatcher-backend/infrastructure/endpoint/amqp/src/main/java/pt/sensae/services/notification/dispatcher/backend/infrastructure/endpoint/amqp/mapper/AddresseeMapper.java
package pt.sensae.services.notification.dispatcher.backend.infrastructure.endpoint.amqp.mapper;

import pt.sensae.services.notification.dispatcher.backend.domain.adressee.Addressee;
import pt.sensae.services.notification.dispatcher.backend.domain.adressee.AddresseeConfig;
import pt.sensae.services.notification.dispatcher.backend.domain.adressee.AddresseeId;
import pt.sensae.services.notification.dispatcher.backend.domain.adressee.DeliveryType;
import pt.sensae.services.notification.dispatcher.backend.domain.contentType.ContentType;
import pt.sensae.services.notification.dispatcher.backend.domain.notification.NotificationLevel;
import pt.sensae.services.notification.dispatcher.backend.infrastructure.endpoint.amqp.model.AddresseeDTO;
import pt.sensae.services.notification.dispatcher.backend.infrastructure.endpoint.amqp.model.DeliveryTypeDTO;
import pt.sensae.services.notification.dispatcher.backend.infrastructure.endpoint.amqp.model.NotificationLevelDTO;

import java.util.UUID;
import java.util.stream.Collectors;

public class AddresseeMapper {

    public static Addressee dtoToDomain(AddresseeDTO addressee) {
        var id = new AddresseeId(UUID.fromString(addressee.id));
        var entries = addressee.entries.stream().map(entry -> {
            var contentType = new ContentType(entry.category, entry.subCategory, extract(entry.severity));
            return new AddresseeConfig(contentType, extract(entry.deliveryType), entry.mute);
        }).collect(Collectors.toSet());
        return Addressee.of(id, entries);
    }

    private static NotificationLevel extract(NotificationLevelDTO dto) {
        return switch (dto) {
            case WATCH -> NotificationLevel.WATCH;
            case INFORMATION -> NotificationLevel.INFORMATION;
            case ADVISORY -> NotificationLevel.ADVISORY;
            case WARNING -> NotificationLevel.WARNING;
            case CRITICAL -> NotificationLevel.CRITICAL;
        };
    }

    private static DeliveryType extract(DeliveryTypeDTO dto) {
        return switch (dto) {
            case EMAIL -> DeliveryType.EMAIL;
            case NOTIFICATION -> DeliveryType.NOTIFICATION;
            case SMS -> DeliveryType.SMS;
        };
    }
}

File: project/frontend-services/libs/notification-management/mapper/src/lib/mappers/AddresseeMapper.ts
import {
  AddresseeConfiguration,
  AddresseeConfigurationEntry,
  DeliveryType
} from "@frontend-services/notification-management/model";
import {AddresseeDTO, DeliveryTypeDTO} from "@frontend-services/notification-management/dto";
import {ContentTypeMapper} from "./ContentTypeMapper";

export class AddresseeMapper {
  static dtoToModel(dto: AddresseeDTO): AddresseeConfiguration {
    const entries = dto.entries.map(conf => new AddresseeConfigurationEntry(this.dtoToModelDeliveryType(conf.deliveryType), ContentTypeMapper.dtoToModel(conf.contentType), conf.mute));
    return new AddresseeConfiguration(entries);
  }

  static modelToDto(model: AddresseeConfiguration): AddresseeDTO {
    const entries = model.entries.map(conf => ({
      mute: conf.mute,
      deliveryType: this.modelToDtoDeliveryType(conf.deliveryType),
      contentType: ContentTypeMapper.modelToDto(conf.contentType)
    }));
    return {entries};
  }

  private static modelToDtoDeliveryType(model: DeliveryType): DeliveryTypeDTO {
    switch (model) {
      case DeliveryType.NOTIFICATION:
        return DeliveryTypeDTO.NOTIFICATION;
      case DeliveryType.SMS:
        return DeliveryTypeDTO.SMS;
      case DeliveryType.EMAIL:
        return DeliveryTypeDTO.EMAIL;
      case DeliveryType.UI:
        return DeliveryTypeDTO.UI;
    }
  }

  private static dtoToModelDeliveryType(dto: DeliveryTypeDTO): DeliveryType {
    switch (dto) {
      case DeliveryTypeDTO.NOTIFICATION:
        return DeliveryType.NOTIFICATION;
      case DeliveryTypeDTO.SMS:
        return DeliveryType.SMS;
      case DeliveryTypeDTO.EMAIL:
        return DeliveryType.EMAIL;
      case DeliveryTypeDTO.UI:
        return DeliveryType.UI;
    }
  }
}
